File: orcastork/datapoints/collection.py
"""``DataPointSet`` — a keyed collection with merge-on-add (the dedup primitive).

Identity is ``(type, value)`` (timestamps excluded), so adding an equal DataPoint
**merges** into the existing entry — keeping ``first_retrieved`` and advancing
``last_retrieved`` — rather than duplicating or dropping the new timestamp. The
in-memory store and delta machinery build on this.
"""

from __future__ import annotations

from collections.abc import Iterable, Iterator
from dataclasses import dataclass
from enum import Enum
from typing import Any

from .base import BaseDataPoint, identity_key


class MergeKind(Enum):
    ADDED = 'added'  # a new (type, value) identity
    UPDATED = 'updated'  # an existing identity re-observed (last_retrieved bumped)


@dataclass(frozen=True)
class MergeResult:
    kind: MergeKind
    data_point: BaseDataPoint[Any]


class DataPointSet:
    def __init__(self, items: Iterable[BaseDataPoint[Any]] = ()) -> None:
        self._items: dict[tuple[str, Any], BaseDataPoint[Any]] = {}
        for item in items:
            self.add(item)

    @staticmethod
    def _key(data_point: BaseDataPoint[Any]) -> tuple[str, Any]:
        return identity_key(data_point)

    def add(self, data_point: BaseDataPoint[Any]) -> MergeResult:
        """Insert a new identity, or merge into an existing one (bumping ``last_retrieved``)."""
        key = self._key(data_point)
        existing = self._items.get(key)
        if existing is None:
            self._items[key] = data_point
            return MergeResult(MergeKind.ADDED, data_point)
        merged = existing.reobserved(data_point.last_retrieved)
        self._items[key] = merged
        return MergeResult(MergeKind.UPDATED, merged)

    def all(self) -> tuple[BaseDataPoint[Any], ...]:
        return tuple(self._items.values())

    def __iter__(self) -> Iterator[BaseDataPoint[Any]]:
        return iter(self._items.values())

    def __len__(self) -> int:
        return len(self._items)

    def __contains__(self, data_point: object) -> bool:
        return isinstance(data_point, BaseDataPoint) and self._key(data_point) in self._items

File: typescript/src/orcastork/datapoints/collection.ts
/**
 * `DataPointSet` — a keyed collection with merge-on-add (the dedup primitive).
 *
 * Identity is `(type, value)` (timestamps excluded), so adding an equal DataPoint **merges** into
 * the existing entry — keeping `firstRetrieved` and advancing `lastRetrieved` — rather than
 * duplicating or dropping the new timestamp. The in-memory store and delta machinery build on this.
 *
 * @module
 */

import type { AnyDataPoint } from './base.js';
import { BaseDataPoint, identityKey } from './base.js';

/** How a DataPoint landed: a new `(type, value)` identity, or a re-observation of one held. */
export const MergeKind = {
  /** A new `(type, value)` identity. */
  ADDED: 'added',

  /** An existing identity re-observed (`lastRetrieved` bumped). */
  UPDATED: 'updated',
} as const;

/** One of the two {@link MergeKind} values; the strings reach the audit trail and the wire. */
export type MergeKind = (typeof MergeKind)[keyof typeof MergeKind];

/** What one `add` did, and the entry the set holds afterwards. */
export interface MergeResult {
  readonly kind: MergeKind;

  readonly dataPoint: AnyDataPoint;
}

/** A keyed set of DataPoints: adding an identity already present merges into it. */
export class DataPointSet {
  private readonly items = new Map<string, AnyDataPoint>();

  public constructor(items: Iterable<AnyDataPoint> = []) {
    for (const item of items) {
      this.add(item);
    }
  }

  /** Insert a new identity, or merge into an existing one (bumping `lastRetrieved`). */
  public add(dataPoint: AnyDataPoint): MergeResult {
    const key = identityKey(dataPoint);
    const existing = this.items.get(key);
    if (existing === undefined) {
      this.items.set(key, dataPoint);
      return { kind: MergeKind.ADDED, dataPoint };
    }
    const merged = existing.reobserved(dataPoint.lastRetrieved);
    this.items.set(key, merged);
    return { kind: MergeKind.UPDATED, dataPoint: merged };
  }

  /** Every held DataPoint, in first-insertion order. */
  public all(): readonly AnyDataPoint[] {
    return [...this.items.values()];
  }

  /**
   * Whether this identity is held — timestamps excluded.
   *
   * Anything that is not a DataPoint is simply absent, rather than an error: the guard is what
   * keeps a membership test on a foreign value from blowing up inside the identity normalization.
   */
  public has(dataPoint: unknown): boolean {
    return dataPoint instanceof BaseDataPoint && this.items.has(identityKey(dataPoint));
  }

  /** How many identities the set holds. */
  public get size(): number {
    return this.items.size;
  }

  public [Symbol.iterator](): Iterator<AnyDataPoint> {
    return this.items.values();
  }
}
